packages/amplify-graphql-function-transformer/src/graphql-function-transformer.ts

```diff
@@ -195,3 +255,256 @@ const lambdaArnKey = (name: string, region?: string, accountId?: string): string
+  // Now we validate that this field has the shape 'success: Boolean!'
+  const schemaDefinedTypeHasValidShape =
+    expectedField.name.value === 'success' &&
+    expectedField.type.kind === Kind.NON_NULL_TYPE &&
+    expectedField.type.type.kind === Kind.NAMED_TYPE &&
+    expectedField.type.type.name.value === 'Boolean';
+
+  if (!schemaDefinedTypeHasValidShape) {
+    throw new InvalidDirectiveError(errorMessage);
+  }
+};
+
+/**
+ * Validates that the parent type is valid for 'invocationType: Event'
+ * - parent type === 'Query' | 'Mutation'
+ * This is done to ensure we don't accidentally add Event invocation type support to other types, e.g. `@model` or subscriptions.
+ * @param config the {@link FunctionDirectiveConfiguration} for 'invocationType: Event' definitions.
+ */
+const validateIsSupportedEventInvocationParentType = (config: FunctionDirectiveConfiguration): void => {
+  const isValidParentType = config.resolverTypeName === 'Query' || config.resolverTypeName === 'Mutation';
+
+  if (!isValidParentType) {
+    throw new InvalidDirectiveError("@function definition with 'invocationType: Event' must be defined on Query or Mutation field.");
+  }
+};
+
+// #endregion Validation helpers
+
+// #region Debug Description Helpers
+
+/**
+ * Use this to generate a description of a type for contextual error messages.
+ * @param typeNode the {@link TypeNode} to describe.
+ * @returns a textual description of the {@link TypeNode} in `string` form.
+ */
+const typeDebugDescription = (typeNode: TypeNode): string => {
+  /*
+  Int -- NamedType
+  Int! -- NonNullType.NamedType -- '' / !
+  [Int] -- ListType.NamedType -- [ / ]
+  [Int]! -- NonNullType.ListType.NamedType -- '' / ! --> [ / ]!
+  [Int!] -- ListType.NonNullType.NamedType -- [ / ] --> [ / !]
+  [Int!]! -- NonNullType.ListType.NonNullType.NamedType -- '' / ! --> [ / ]! --> [ / !]!
+  */
+  // eslint-disable-next-line consistent-return
+  const description = (node: TypeNode, prefix = '', suffix = ''): string => {
+    switch (node.kind) {
+      case Kind.LIST_TYPE:
+        return description(node.type, prefix + '[', ']' + suffix);
+      case Kind.NON_NULL_TYPE:
+        return description(node.type, prefix, '!' + suffix);
+      case Kind.NAMED_TYPE:
+        return `${prefix}${node.name.value}${suffix}`;
+    }
+  };
+
+  return description(typeNode);
+};
+
+/**
+ * Use this to generate a field description for contextual error messages in the form of:
+ * `field-name(optional-args): return-type `@function`(name: name, invocationType: invocation-type)`
+ * @param config the {@link FunctionDirectiveConfiguration} of the `@function` directive.
+ * @param field the {@link FieldDefinitionNode} on which the `@function` directive is defined.
+ * @returns a description of the field with `@function` directive appropriate for use in error messages.
+ */
+const fieldDebugDescription = (config: FunctionDirectiveConfiguration, field: FieldDefinitionNode): string => {
```

Comment:
We may want to consider adding this to `amplify-graphql-transformer-core/utils`. If so, I'll take that on in a follow up.